flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/BufferConsumer.java

```diff
@@ -44,6 +44,9 @@
 
 	private int currentReaderPosition;
 
+	/** Whether this BufferConsumer is copied from another BufferConsumer instance. */
+	private final boolean isCopied;
```

Comment:
TBH I do not like the way of introducing this property inside `BufferConsumer` for several concerns:

1. This property is not the basic characters of buffer or orienting specific scenarios. And it is somehow orienting/relying on the implementation details.

2. In essence this property is coupled with both broadcast improvement and compression limitation. In other words it is not easy to explain why one `BufferConsumer` would be copied by others to make others understand the background. Firstly we have to explain that it is caused by the specific broadcast improvement, and further we have to explain the current compression implementation has the limitation of reusing the original buffer to copy compressed results only for pipelined partition.

Therefore I prefer to introducing the compression feature only for batch jobs ATM, and our initial motivation was also for reducing the disk IO cost. Although for streaming job the compression can also reduce the network bandwidth, it is meanwhile bringing the extra CPU cost and compression/decompression time cost. So for the streaming job without bottleneck of network bandwidth, it is not suggested to enable this feature.

Most importantly the current compression implementation for streaming job brings more customize changes, both for this property in `BufferConsumer` and also the local channel property while getting next buffer from view. I guess we might have better solutions for streaming compression future. WDYT @pnowojski ?

Reply:
I think I'm not that concerned about the issues that you mentioned. I don't think this property is tightly coupled with anything else, it just tells whether this `BufferConsumer` was copied or not. If anything, the issue is that compression happens "in place" modifing the data, which is the source of the problem. If we already agreed to such "hack" (for valid reasons), I think it is ok to for compressor to check whether the `BufferConsumer` was copied or not.

However I have a different concern. If we are copying the `BufferConsumer`, the first one will not have this flag set. Is this an issue? Currently it might work, because of the following loop:
```
try (BufferConsumer bufferConsumer = builder.createBufferConsumer()) {
	for (int channel = 0; channel < numberOfChannels; channel++) {
		targetPartition.addBufferConsumer(bufferConsumer.copy(), channel);
	}
}
```
but it could be also implemented as follows:
```
for (int channel = 0; channel < numberOfChannels - 1; channel++) {
	targetPartition.addBufferConsumer(bufferConsumer.copy(), channel);
}
targetPartition.addBufferConsumer(bufferConsumer, numberOfChannels - 1);
```

Reply:
> If we are copying the BufferConsumer, the first one will not have this flag set. Is this an issue?

Exactly, if we agree to rely on this way for handling the special broadcast case, we should tag this flag as true not only in the `BufferConsumer#copy()`, but also make sure all the `BufferConsumer` instances tagged if they are sharing the same `BufferBuilder`.

Also we might need to refactor this field renaming. It should indicate whether there are multiple `BufferConsumer` instances share the same `BufferBuilder`, not indicating for neither it is copied by others nor it is copying from others.

Reply:
I guess we can also support compression for broadcast mode in streaming job via an more elegant way if have more time to think through, then we can avoid bringing this tag field in `BufferConsumer`. That is why I suggested supporting the compression only for batch jobs for nearly release, and make it also ready for streaming job in future.

But I am not strongly objecting the current way if have no concerns for it. @pnowojski

Reply:
Thanks for the review and comments. I will fix the problem and update the PR soon.

Reply:
> if have more time to think through, then we can avoid bringing this tag field in BufferConsumer. 

I think we can always refactor it in the future, if there will be such need.

>  That is why I suggested supporting the compression only for batch jobs for nearly release, and make it also ready for streaming job in future.

As discussed offline, I think the only argument to drop pipelined compression for now would be, if we do not want the pipelined compression happen on this level at all. For example if we wanted to just just Netty's built-in compression. If we release the code as it is now, then if we will change our minds to re-implement pipelined compression in netty, we would brake the config options compatibility.

All in all, I think the cost of the current implementation, with the existing limitation "compression not working for broadcasting", is relatively low, so I would be inclined to keep it more or less as it is.

The one change that I would request is to fix this "first buffer consumer issue", by changing the `isCopied` flag, to `isCopyable`, which would be set to "true/false" even for the first `BufferConsumer`. If the `BufferConsumer` is copyable, we can not compress it.